processing/src/main/java/io/druid/segment/StringDimensionQueryHelper.java

```diff
@@ -0,0 +1,443 @@
+        values[i] = value != null ? value : CardinalityAggregator.NULL_STRING;
+      }
+      // Values need to be sorted to ensure consistent multi-value ordering across different segments
+      Arrays.sort(values);
+      for (int i = 0; i < size; ++i) {
+        if (i != 0) {
+          hasher.putChar(CardinalityAggregator.SEPARATOR);
+        }
+        hasher.putUnencodedChars(values[i]);
+      }
+    }
+  }
+
+  public void hashValues(Object dimSelector, HyperLogLogCollector collector)
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    for (final Integer index : selector.getRow()) {
+      final String value = selector.lookupName(index);
+      collector.add(CardinalityAggregator.hashFn.hashUnencodedChars(value == null ? CardinalityAggregator.NULL_STRING : value).asBytes());
+    }
+  }
+
+  @Override
+  public int getGroupingKeySize()
+  {
+    return Ints.BYTES;
+  }
+
+  @Override
+  public Comparator<byte[]> getGroupingKeyByteComparator()
+  {
+    return GROUPING_KEY_COMPARATOR;
+  }
+
+  @Override
+  public void readDimValueFromGroupingKey(
+      Map<String, Object> theEvent, String outputName, Object dimSelector, ByteBuffer keyBuffer
+  )
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    final int dimVal = keyBuffer.getInt();
+    if (dimVal != GROUP_BY_MISSING_VALUE) {
+      theEvent.put(outputName, selector.lookupName(dimVal));
+    }
+  }
+
+  @Override
+  public List<ByteBuffer> addDimValuesToGroupingKey(
+      Object selector, ByteBuffer key, Function<ByteBuffer, List<ByteBuffer>> updateValuesFn
+  )
+  {
+    List<ByteBuffer> unaggregatedBuffers = null;
+    final DimensionSelector dimSelector = (DimensionSelector) selector;
+    final IndexedInts row = dimSelector.getRow();
+    if (row == null || row.size() == 0) {
+      ByteBuffer newKey = key.duplicate();
+      newKey.putInt(GROUP_BY_MISSING_VALUE);
+      unaggregatedBuffers = updateValuesFn.apply(newKey);
+    } else {
+      for (Integer dimValue : row) {
+        ByteBuffer newKey = key.duplicate();
+        newKey.putInt(dimValue);
+        unaggregatedBuffers = updateValuesFn.apply(newKey);
+      }
+    }
+    return unaggregatedBuffers;
+  }
+
+  @Override
+  public Object getRowFromDimSelector(Object dimSelector) {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    IndexedInts values = selector == null ? EmptyIndexedInts.EMPTY_INDEXED_INTS : selector.getRow();
+    return values;
+  }
+
+  @Override
+  public int initializeGroupingKeyV2Dimension(
+      final Object valuesObj,
+      final ByteBuffer keyBuffer,
+      final int keyBufferPosition
+  )
+  {
+    IndexedInts values = (IndexedInts) valuesObj;
+    final int rowSize = values.size();
+    if (rowSize == 0) {
+      keyBuffer.putInt(keyBufferPosition, GROUP_BY_MISSING_VALUE);
+    } else {
+      keyBuffer.putInt(keyBufferPosition, values.get(0));
```

Comment:
Documentation to this method says "Read the first value within a row values object and _add_ that value to the keyBuffer at keyBufferPosition", but this implementation overwrites the previous value.